### PR TITLE
refactor(storage): refactor bloom index to use vectorized siphash function

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

refactor bloom index to use vectorized `siphash` function

The time-consuming comparison of inserting int values in my local computer is as follows:

rows     | before  | after
-------- | ------- | -------
1000     | 0.059s  | 0.034s
10000    | 0.108s  | 0.068s
100000   | 0.263s  | 0.183s
1000000  | 2.092s  | 1.338s
10000000 | 26.349s | 21.486s

part of #9480

### CODE REVIEW

File: src/query/storages/common/index/src/bloom.rs [hunk 1]
```diff
@@ -125,42 +137,100 @@ impl BlockFilter {
...
+                let scalar = PhysicalScalar::Function {
+                    name: "siphash".to_string(),
+                    params: vec![],
+                    args: vec![arg],
+                    return_type,
+                };
+                let expr = scalar.as_expr()?;
+                exprs.push(expr);
+
+                let source_columns = blocks
```
Comment (sundy-li) [inline src/query/storages/common/index/src/bloom.rs:169]: `concat` already cover the convert full column codes.

File: src/query/storages/common/index/src/bloom.rs [hunk 2]
```diff
@@ -125,42 +137,100 @@ impl BlockFilter {
...
+                .run(expr)
+                .map_err(|(_, e)| ErrorCode::Internal(format!("eval siphash failed: {}.", e)))?;
+            let col = value.convert_to_full_column(field.data_type(), data_block.num_rows());
+            let (column, validity) = if field.data_type().is_nullable() {
+                let nullable_column =
+                    NullableType::<UInt64Type>::try_downcast_column(&col).unwrap();
+                (nullable_column.column, nullable_column.validity)
+            } else {
+                let column = UInt64Type::try_downcast_column(&col).unwrap();
+                let validity = constant_bitmap(true, column.len()).into();
```
Comment (sundy-li) [inline src/query/storages/common/index/src/bloom.rs:198]: extra allocation for non-nullable types.

File: src/query/storages/common/index/src/bloom.rs [hunk 3]
```diff
@@ -125,42 +137,100 @@ impl BlockFilter {
...
+                (column, validity)
+            };
+
+            // create filter per column
+            let mut filter_builder = Xor8Builder::create();
+            let filter = if validity.unset_bits() > 0 {
```
Comment (sundy-li) [inline src/query/storages/common/index/src/bloom.rs:204]: `validity.is_none() || validity.map... `

File: src/query/storages/common/index/src/bloom.rs [hunk 4]
```diff
@@ -261,7 +331,30 @@ impl BlockFilter {
...
+                };
+                let scalar = PhysicalScalar::Function {
+                    name: "siphash".to_string(),
+                    params: vec![],
+                    args: vec![arg],
+                    return_type: DataType::Number(NumberDataType::UInt64),
+                };
+                let expr = scalar.as_expr()?;
+
+                let data_block = DataBlock::empty();
```
Comment (sundy-li) [inline src/query/storages/common/index/src/bloom.rs:349]: num rows should be 1 ?

File: src/query/storages/common/index/src/bloom.rs [hunk 5]
```diff
@@ -261,7 +331,30 @@ impl BlockFilter {
...
+                let evaluator = Evaluator::new(&data_block, self.func_ctx, &BUILTIN_FUNCTIONS);
+                let value = evaluator.run(&expr).map_err(|(_, e)| {
+                    ErrorCode::Internal(format!("eval siphash failed: {}.", e))
+                })?;
+                let number_scalar = value.into_scalar().unwrap().into_number().unwrap();
```
Comment (sundy-li) [inline src/query/storages/common/index/src/bloom.rs:354]: need to do it in try_create function.

```
hashs: HashMap<Scalar, Enum { DataValue,  u64 } >
```

**Comment by BohuTANG** [comment]
After changing, what's improved? Can not see it from the PR summary.

File: src/query/storages/common/index/src/bloom.rs (hunk 1 above)
Comment (b41sh) [reply src/query/storages/common/index/src/bloom.rs:169]: use `DataBlock::concat` ?

File: src/query/storages/common/index/src/bloom.rs (hunk 4 above)
Comment (b41sh) [reply src/query/storages/common/index/src/bloom.rs:349]: empty block is ok

**Comment by sundy-li** [comment]
I think the insert performance can be compared.

**Comment by b41sh** [comment]
> After changing, what's improved? Can not see it from the PR summary.

@BohuTANG @sundy-li  I have add some performance comparison of insert values, the time consumed to insert 10000000 rows of int values has been reduced by about 5s.

**Comment by BohuTANG** [comment]
> > After changing, what's improved? Can not see it from the PR summary.
> 
> @BohuTANG @sundy-li I have add some performance comparison of insert values, the time consumed to insert 10000000 rows of int values has been reduced by about 5s.

Can we give some performance numbers for it? T

**Comment by BohuTANG** [comment]
@mergify update

File: src/query/storages/common/index/src/bloom.rs [hunk 6]
```diff
@@ -199,22 +199,22 @@ impl BlockFilter {
...
-                        if !b {
-                            digests.push(0);
-                        } else {
-                            digests.push(v);
-                        }
-                    });
-                filter_builder.build_from_digests(&digests)?
+                let mut column_iter = column.deref().iter();
+                let mut validity_iter = validity.into_iter();
+                filter_builder.add_digests(std::iter::from_fn(move || {
```
Comment (sundy-li) [inline src/query/storages/common/index/src/bloom.rs:206]: why not:
```
let it = UInt64Type::iter_column(&column)
                    .zip(validity.iter())
                    .for_each(|(v, b)| {
                        if !b {
                            &0
                        } else {
                             &v
                        }
                    });
                    
    filter_builder.add_digests(it);               
```

File: src/query/storages/common/index/src/bloom.rs [hunk 7]
```diff
@@ -220,11 +293,41 @@ impl BlockFilter {
...
+            args: vec![arg],
+            return_type: DataType::Number(NumberDataType::UInt64),
+        };
+        let expr = siphash_func.as_expr()?;
+
+        let data_block = DataBlock::empty();
+        let evaluator = Evaluator::new(&data_block, func_ctx, &BUILTIN_FUNCTIONS);
+        let value = evaluator
+            .run(&expr)
+            .map_err(|(_, e)| ErrorCode::Internal(format!("eval siphash failed: {}.", e)))?;
```
Comment (andylokandy) [inline src/query/storages/common/index/src/bloom.rs:318]: You can use `common_expression::eval_function()` instead.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**